### ISSUE TITLE
Elements in Attributes and Member Elements

### ISSUE THREAD

**Issue by smizell** [refractproject/refract-spec#12]
## Problem

There is a need to have full elements inside the attributes hash structure. For example:

``` json
["foo", {}, { "bar": ["baz", {}, {}, null] }, null]
```

Here we have a `bar` attribute that has a `baz` element as it's value. The current issue that we've come across is that when the attributes are "refracted," we get the following:

``` json
["foo", {}, [
  ["member", {}, {}, {
    "key": ["string", {}, {}, "baz"],
    "value": ["array", {}, {}, [
      ["string", {}, {}, "baz"],
      ["object", {}, {}, null],
      ["object", {}, {}, null],
      ["null", {}, {}, null]
    ]]
  }]
], null]
```

This is undesired because it refracts an element that should not be refracted. The desired outcome is:

``` json
["foo", {}, [
  ["member", {}, {}, {
    "key": ["string", {}, {}, "baz"],
    "value": ["baz", {}, {}, null]
  }]
], null]
```
## Potential Solution

I think one way we can solve how to handle this is by relying on the element definition to determine how the attributes and content should be refracted. Simply refracting an attributes object/hash really gives no benefit on its own unless you have some understanding of the semantics of the attributes. Therefore, we can allow elements to exist alongside objects/hashes so long as the definition is used in conjunction.

For example, here is a definition of a `foo` element in MSON.

``` markdown
# Foo (Element)

- `element` foo (string, fixed)
- `attributes`
    - `bar` (string)
```

Given this definition, we know this instance is valid.

``` json
["foo", {}, { "bar": "baz" }, null]
```

Since the `baz` value is a string, it can also be refracted on its own.

``` json
["foo", {}, { "bar": ["string", {}, {}, "baz"] }, null]
```

Additionally, there is no reason you can't swap out the `string` element here for another element that extends the primitive `string`.

``` json
["foo", {}, { "bar": ["abc", {}, {}, "baz"] }, null]
```

In relying on the definitions, we can have primitive or refracted values mixed in and a parser library should be able to differentiate quite easily. In addition to this, you MAY refract attributes in its entirety.

``` json
["foo", {}, [
  ["member", {}, {}, {
    "key": ["string", {}, {}, "bar"],
    "value": ["string", {}, {}, "baz"]
  }]
], null]
```

This will need to come with a constraint, which is that objects may not be refracted and mixed into normal hashes in full refract. Same goes for arrays in compact refract. This is because you can't tell the difference between the two in their given forms.
## Side Benefit

If we do this, we can greatly simplify member elements. We can then define a member as such:

``` markdown
# Member (Element)

- `element` member (string, fixed)
- `attributes`
    - `name` (string)
```

Now, given an object like this:

``` json
{
  "foo": "bar"
}
```

It may be refracted like this:

``` json
["object", {}, {}, [
  ["member", {}, { "name": "foo" },
    ["string", {}, {}, "bar"]]
]]
```

If it is desired to add additional data about the actual `foo` string, you can refract it.

``` json
["object", {}, {}, [
  ["member", {}, { "name": ["string", {}, {}, "foo"] },
    ["string", {}, {}, "bar"]]
]]
```

The reason we did not go this route at first was because of the issue with mixing hashes and elements, but if solved by relying on the definition, we can go back to simplify this like this.

cc: @fosrias, @zdne @kylef

**Comment by smizell** [comment]
Though it may be implied here, I just wanted to add to this that relying on the element definition means you can define attributes as elements themselves. I think if we follow this (which is maybe what everyone was wanting to do in the first place), we can get away with this by not trying to refract attributes without any given knowledge of the definition of the element.

**Comment by zdne** [comment]
> I think one way we can solve how to handle this is by relying on the element definition to determine how the attributes and content should be refracted

:+1: I love you @smizell ! Let's do it!

**Comment by zdne** [comment]
@smizell do we need to change anything in the base spec? If so, will you please take care of it when your schedule allows?

**Comment by smizell** [comment]
@zdne I would like to clarify that more, because how you refract attributes and meta is mostly not existent in the spec.

Additionally, if we go this route, what are your thoughts on modifying/simplifying the member element?

**Comment by zdne** [comment]
> thoughts on modifying/simplifying the member element?

I am all up for it! The question is whether that `name` should be in `meta` or in `attributes` ?

**Comment by fosrias** [comment]
@zdne @smizell Interesting thought about `name`. What if we only put in `meta` what refract needs to do its own business and if something needs a `name` or `title` attribute, that shows up in the namespace.

That way there is not a collision in usage of `name` or whatever in `meta`.

**Comment by zdne** [comment]
@fosrias are you saying an element title or description should be in attributes – not in meta?

**Comment by fosrias** [comment]
@zdne Just wondering about minimal set that refract really needs there. It may be better to leave those there, I am just asking the question in principal.

**Comment by smizell** [comment]
I actually had `name` in the `attributes` of the member element, if that makes a difference to the discussion here. I don't think `name` should be included in `meta`.

**Comment by zdne** [comment]
@fosrias @smizell well I do not like `meta` in the first place, to me its exactly bringing just these problems, but that is another discussion I guess. For the time being I would stick what we have and let's just say that Refract base spec elements are the only ones privileged to have something in `meta`

**Comment by zdne** [comment]
> I actually had name in the attributes of the member element, if that makes a difference to the discussion here. I don't think name should be included in meta

Wasn't the whole point of `meta` to "namespace-out" / separate refract base specific attributes from `attributes`?

**Comment by smizell** [comment]
> Wasn't the whole point of meta to "namespace-out" / separate refract base specific attributes from attributes?

Yes, but `name` is specific only to the `member` element. We can use the attributes how we want for specific elements. Using meta was for attributes that transcend all elements and aren't really element-specific. Rather, there are meta things like a unique ID or classification.

> I do not like meta in the first place

In light of what I said above, could we safely and easily combine attributes and meta? If so, it's a quick change. Additionally, I had initially namespaced these things within attributes.
- `element` foo
- `attributes` (object)
  - `meta` (object)
    - `id` bar
- `content` baz (string)

**Comment by zdne** [comment]
@smizell I would very much prefer to joining these two together, if still possible, that is

**Comment by zdne** [comment]
I believe this was resolved in #14
